extensions/socket/internal.m

```diff
@@ -0,0 +1,362 @@
+    [asyncSocket disconnect];
+
+    [skin pushNSObject:asyncSocket];
+    return 1;
+}
+
+/// hs.socket:info() -> table
+/// Method
+/// Returns information on the socket instance
+///
+/// Parameters:
+///  * None
+///
+/// Returns:
+///  * A table containing the following keys:
+///   * connectedHost
+///   * connectedPort
+///   * localHost
+///   * localPort
+///   * isConnected
+///   * isIPv4
+///   * isIPv6
+///   * isIPv4Enabled
+///   * isIPv6Enabled
+///   * isIPv4PreferredOverIPv6
+///   * isSecure
+///
+static int socket_info(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TBREAK];
+
+    HSAsyncSocket*  asyncSocket = [skin luaObjectAtIndex:1 toClass:"HSAsyncSocket"];
+
+    NSString *connectedHost = [asyncSocket connectedHost];
+    NSNumber *connectedPort = [NSNumber numberWithUnsignedShort:[asyncSocket connectedPort]];
+    NSString *localHost = [asyncSocket localHost];
+    NSNumber *localPort = [NSNumber numberWithUnsignedShort:[asyncSocket localPort]];
+    NSNumber *isConnected = [NSNumber numberWithBool:[asyncSocket isConnected]];
+    NSNumber *isIPv4 = [NSNumber numberWithBool:[asyncSocket isIPv4]];
+    NSNumber *isIPv6 = [NSNumber numberWithBool:[asyncSocket isIPv6]];
+    NSNumber *isIPv4Enabled = [NSNumber numberWithBool:[asyncSocket isIPv4Enabled]];
+    NSNumber *isIPv6Enabled = [NSNumber numberWithBool:[asyncSocket isIPv6Enabled]];
+    NSNumber *isIPv4PreferredOverIPv6 = [NSNumber numberWithBool:[asyncSocket isIPv4PreferredOverIPv6]];
+    NSNumber *isSecure = [NSNumber numberWithBool:[asyncSocket isSecure]];
+
+    connectedHost = connectedHost ? connectedHost : @"";
+    localHost = localHost ? localHost : @"";
+
+    NSDictionary *info = @{
+        @"connectedHost" : connectedHost,
+        @"connectedPort" : connectedPort,
+        @"localHost" : localHost,
+        @"localPort" : localPort,
+        @"isConnected": isConnected,
+        @"isIPv4": isIPv4,
+        @"isIPv6": isIPv6,
+        @"isIPv4Enabled": isIPv4Enabled,
+        @"isIPv6Enabled": isIPv6Enabled,
+        @"isIPv4PreferredOverIPv6": isIPv4PreferredOverIPv6,
+        @"isSecure": isSecure,
+    };
+
+    [skin pushNSObject:info];
+    return 1;
+}
+
+// Pushes the provided HSAsyncSocket onto the Lua Stack as a hs.socket userdata object
+static int socket_tolua(lua_State *L, id obj) {
+    HSAsyncSocket *socket = obj;
+    void** ptr = lua_newuserdata(L, sizeof(HSAsyncSocket *));
+    *ptr = (__bridge_retained void *)socket;
+    luaL_getmetatable(L, USERDATA_TAG);
+    lua_setmetatable(L, -2);
+    return 1;
+}
+
+static id lua_tosocket(lua_State *L, int idx) {
+    void *ptr = luaL_testudata(L, idx, USERDATA_TAG);
+    if (ptr) {
+        return (__bridge HSAsyncSocket *)*((void **)ptr);
+    } else {
+        return nil;
+    }
+}
+
+static int socket_objectGC(lua_State *L) {
+    HSAsyncSocket *asyncSocket = [[LuaSkin shared] luaObjectAtIndex:1 toClass:"HSAsyncSocket"];
```

Comment:
Note that, however the userdata stuff is handled, there needs to be a (__bridge_transfer HSAsyncSocket *) cast somewhere to balance out the __bridge_retained. In this current iteration (and probably most others), it would be here in the __gc function.